proposals/distributed-async-epoch-reward.md

```diff
@@ -0,0 +1,220 @@
+process into two distinct phases: (1) rewards calculation, and (2) rewards
+distribution.
+
+
+### Rewards Calculation
+
+In contrast to the current method of computing epoch rewards, which involves
+directly computing rewards at the first slot of the epoch boundary in the bank
+runtime, the new approach spreads the reward computation over multiple blocks.
+
+As an example, in the Solana Labs validator client, the reward computation is
+moved to a background service known as the `EpochRewardCalculationService`.
+This service starts automatically when a validator starts and listens to a
+designated channel for any incoming rewards calculation requests, performing
+the necessary rewards calculations accordingly.
+
+When a block first crosses the epoch boundary, its bank will send a request to
+the `EpochRewardCalculationService` to start the process of epoch reward
+computation. This marks the start of the reward computation phase.
+
+The reward computation phase lasts for `N` blocks. When reaching block height
+`N` after the start of the `reward calculation phase`, the bank will mark it
+the end the `reward calculation phase`.
+
+At this point, the reward computation result should be available. In case that
+the reward computation result is not available yet at the `N_th_` block-height,
+the bank runtime will block and wait until the computation is finished, before
+processing any blocks with height greater than `N`.
+
+At the end of the reward computation phase (end of block height `N`), a set of
+sysvar accounts will be updated - `EpochRewardProof` and `RewardReserve`.
+
+`EpochRewardProof` sysvar account maintains a history of epoch rewards. The
+internal format of `EpochRewardProof` is an array of entries,
+```
+[(epoch, (reward_amount, root_hash))]
+```
+, where each entry corresponds to the reward for
+one particular epoch. The maximum depth of the history are the same as the
+depth of stake history, which is 512.
+
+When reaching the end of reward computation phase, `EpochRewardProof` are
+updated with the reward computation result. A new entry will be pushed into the
+array. If the length of the history array reaches the maximum depth, the oldest
+entry will popped.
+
+```
+struct RewardHistoryEntry {
+   total: u64,                  // total rewards for this epoch
+   distributed: u64,            // already distributed reward amount
+}
+
+type EpochReward = (Epoch, RewardHistoryEntry);
+
+struct RewardHistory {
+   rewards: [EpochReward; 512],
+}
+```
+
+At the end of reward computation phase, a set of `EpochRewardsReserve` sysvar
+accounts will be updated too. Each `RewardsReserve` records the reward balance
+to be distributed for a particular block at block height `height` in `Reward Distribution`
+phase. The set of stake rewards paid from `EpochRewardsReserve`
+sysvar accounts form a complete partition of the total stake rewards.
+
+```
+struct EpochRewardReserve {
+   balance: u64,
+   height: u64,
```

Comment:
Yes. absolute height. As @mvines suggested, this will make the query for reward distribution per block easier for the user, who many just have the block height.